adapters/repos/db/vector/hnsw/vertex.go

```diff
@@ -0,0 +1,104 @@
+package hnsw
+
+import (
+	"sync"
+)
+
+type vertex struct {
+	id uint64
+	sync.Mutex
+	level       int
+	connections [][]uint64
+	maintenance bool
+}
+
+func (v *vertex) markAsMaintenance() {
```

Comment:
_Performance hint_:
The version below would be more performant in this case since it doesn't incur the overhead of Defer(). 
Defer() is not required in this case because assigning a value to boolean field will not result in any exception.
```
func (v *vertex) markAsMaintenance() {
 	v.Lock()
 	v.maintenance = true
        v.Unlock()
 }
```

Reply:
Same thing applies to `unmarkAsMaintenance() and isUnderMaintenance()`

Reply:
From what I understand there is no longer a performance penalty since [Go 1.14](https://go.dev/doc/go1.14#runtime), the docs explicitly say that defer can be used on the hot path now. I think the compiler just inlines the function call, which basically leads to the same result as rewriting it like this manually.

Reply:
That's almost right quoting Go 1.14
> This release improves the performance of most uses of defer to incur **almost zero overhead** compared to calling the deferred function directly. As a result, defer can now be used in performance-critical code without overhead concerns.

Reply:
But still, even with the almost, the recommendation is now to use it on the hot path:

> As a result, defer can now be used in performance-critical code without overhead concerns.

Happy to change this if you think it's worth it, but we should probably look at the output assembly to see if there is any difference at all.

Reply:
Done.